src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.cs

```diff
@@ -162,12 +177,16 @@ private bool StartNewModelComputation(ICompletionService completionService, Comp
 
             this.sessionOpt = new Session(this, computation, GetCompletionRules(), Presenter.CreateSession(TextView, SubjectBuffer, null));
 
-            var completionProviders = triggerInfo.TriggerReason == CompletionTriggerReason.Snippets
+            var completionProviders = trigger is DisplaySnippetsCompletionTrigger
                 ? GetSnippetCompletionProviders()
                 : GetCompletionProviders();
 
-            sessionOpt.ComputeModel(completionService, triggerInfo, completionProviders, _isDebugger);
-            var filterReason = triggerInfo.TriggerReason == CompletionTriggerReason.BackspaceOrDeleteCommand ? CompletionFilterReason.BackspaceOrDelete : CompletionFilterReason.TypeChar;
+            sessionOpt.ComputeModel(completionService, trigger, completionProviders);
+
+            var filterReason = trigger is BackspaceOrDeleteCharCompletionTrigger
+                ? CompletionFilterReason.BackspaceOrDelete
+                : CompletionFilterReason.TypeChar;
```

Comment:
Doesn't this function also get called during explicit triggers of completion? Is TypeChar the right trigger in this case? Or is this grand unification of "typing forward is the same as explicit trigger" thing I've hoped for?